src/celeritas/ext/GeantImporter.cc

```diff
@@ -481,47 +481,54 @@ std::vector<ImportElement> import_elements()
 /*!
  * Store material-dependent optical properties.
  *
- * This returns:
- * - A vector of optical materials corresponding to an "optical material ID",
- *   and
- * - A "map" (indirection vector) of the optical material ID for each "geometry
- *   material"
+ * This returns a vector of optical materials corresponding to an "optical
+ * material ID".
  */
-std::pair<std::vector<ImportOpticalMaterial>, std::vector<OpticalMaterialId>>
-import_optical()
+std::vector<ImportOpticalMaterial>
+import_optical(detail::GeoOpticalIdMap const& geo_to_opt)
 {
+    if (geo_to_opt.empty())
+    {
+        CELER_LOG(warning)
+            << R"(Optical materials were requested but none are present)";
+        // No optical materials
+        return {};
+    }
+
     auto const& mt = *G4Material::GetMaterialTable();
-    CELER_ASSERT(mt.size() > 0);
+    CELER_ASSERT(mt.size() == geo_to_opt.num_geo());
 
-    std::vector<OpticalMaterialId> geo_to_opt(mt.size());
-    std::vector<ImportOpticalMaterial> result;
+    std::vector<ImportOpticalMaterial> result(geo_to_opt.num_optical());
 
     // Loop over optical materials
-    for (auto mat_idx : range(mt.size()))
+    for (auto geo_mat_id : range(GeoMaterialId{geo_to_opt.num_geo()}))
     {
-        G4Material const* material = mt[mat_idx];
-        CELER_ASSERT(material);
-        CELER_ASSERT(mat_idx == static_cast<std::size_t>(material->GetIndex()));
-
-        // Add optical material properties, if any are present
-        auto const* mpt = material->GetMaterialPropertiesTable();
-        if (!mpt)
+        auto opt_mat_id = geo_to_opt[geo_mat_id];
+        if (!opt_mat_id)
         {
             continue;
         }
+        // Get Geant4 material properties
+        G4Material const* material = mt[geo_mat_id.get()];
+        CELER_ASSERT(material);
+        CELER_ASSERT(geo_mat_id.get()
+                     == static_cast<std::size_t>(material->GetIndex()));
+        auto const* mpt = material->GetMaterialPropertiesTable();
+        CELER_ASSERT(mpt);
 
-        ImportOpticalMaterial optical;
         detail::GeantMaterialPropertyGetter get_property{*mpt};
 
+        // Optical materials should map uniquely
+        ImportOpticalMaterial& optical = result[opt_mat_id.get()];
+        CELER_ASSERT(!optical);
+
         // Save common properties
         bool has_rindex = get_property(&optical.properties.refractive_index,
                                        "RINDEX",
                                        ImportUnits::unitless);
-        if (!has_rindex)
-        {
-            // Refractive index is required: assume the material isn't optical
-            continue;
-        }
+        // Existence of RINDEX should correspond to GeoOpticalIdMap
+        // construction
+        CELER_ASSERT(has_rindex);
```

Comment:
Shouldn't this be a on-all-the-time exception (or at least a warning) to warn the user the wrong type of information is being loaded?

Reply:
No, this truly is just an internal consistency check between the `import_optical` function and the `GeoOpticalIdMap` constructor. The always-on warning at the top of the function is what tells the user if their request is inconsistent with what's there.